vim.validate(): include stacktrace in message

tjdevries opened the pull request:
### No longer exactly old PR.

Now just makes `vim.validate` show a traceback. Doesn't change behavior (still raises error).

This way there are no cryptic messages with no context about who passed the wrong parameters.

Result:

```lua
local function z(val)
  vim.validate { val = { val, 'string' } }
end

local function y(val)
  z(val)
end

local function x(val)
  y(val)
end

x(5)
```

```
Error detected while processing function <SNR>12_save_and_exec:                                                                             
line    6:                                                                                                                                  
E5113: Error while calling lua chunk: /home/tj/test/vim_validate.lua:3: val: expected string, got number                                    
stack traceback:                                                                                                                            
        /home/tj/test/vim_validate.lua:3: in function 'z'                                                                                   
        /home/tj/test/vim_validate.lua:7: in function 'y'                                                                                   
        /home/tj/test/vim_validate.lua:11: in function 'x'                                                                                  
        /home/tj/test/vim_validate.lua:14: in main chunk
```

tjdevries committed:
vim.validate(): include stacktrace in message

tjdevries commented:
@teto do you know how to regenerate the docs when making changes in lua functions?

teto commented:
or instead of calling `error(`, the function could `return false, invalid_key` and have a wrapper erroring out to keep the current behavior. I find the validate function but the forced `error` prevents its reuse.

tjdevries commented:
Ah, like a `is_valid` function that does all the work, and `validate` is jut a function that calls `if_valid` and errors if it is not.

tjdevries commented:
So, currently no way to check that keys of a table follow a certain pattern, right? You'd have to make two calls to validate something like this?

```
t = { point = { 13, 27 } }
```

Might add a way to do that more easily.

teto commented:
I dont think we have that, current validation is limited. Maybe there exists some lua library to handle more complex validation.

tjdevries commented:
@teto does this look good to you? I can add the other validation stuff at a different time if we ever want it.

tjdevries commented:
(if it looks good, I'll just add one or two tests of the `is_valid` and custom error message and then merge later)

justinmk commented:
why do we need a new function?

teto commented:
I would prefer validate to not `error(` as you suggest.

tjdevries commented:
OK, sounds good. I will change it to do that and change all of the `vim.validate` that we have to `assert(vim.validate(...))`

mfussenegger commented:
There was btw also https://github.com/neovim/neovim/pull/11640 which extended the error calls with a traceback.  

assert calls also sound good.  In any case, :+1: for anything that results in errors that point to the callee of validate instead of validate itself

tjdevries commented:
@justinmk can we merge this PR? Anything you want to change about this?

justinmk commented:
going to see if I can modify `pcall_err` so that the `contains()` stuff is avoided.

tjdevries commented:
Awesome, that'd be great. I still haven't fully grokked all the testing stuff, so not attached to anything there.

justinmk committed:
test/vim.validate(): assert normalized stacktrace

- The previous commit lost information in the tests. Instead, add some
  more "normalization" substitutions in pcall_err(), so that the general
  shape of the stacktrace is included in the asserted text.
- Eliminate contains(), it is redundant with matches()

justinmk commented:
pushed a suggestion. also squashed the fixup commits

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -477,48 +477,77 @@ end
+        if not valid then
+          local error_message = string.format("%s: expected %s, got %s", param_name, (spec[3] or '?'), val)
+          if not (optional_message == nil) then
+            error_message = error_message .. string.format(". Info: %s", optional_message)
```
Looks like we don't have a test covering this

justinmk commented on test/functional/lua/vim_spec.lua:
```diff
@@ -579,19 +598,19 @@ describe('lua stdlib', function()
     ]]))
 
-    eq('Error executing lua: .../shared.lua: invalid "behavior": nil',
+    eq('.../helpers.lua:0: Error executing lua: shared.lua:0: invalid "behavior": nil',
```
- It's kinda ugly that helpers.lua is here, but that's what is actually happening so no point in hiding it. 
- The `.../shared.lua` was wrong, there was no full path to `shared.lua`. The new form is correct.

tjdevries commented:
@justinmk OK, I think I fixed all your comments and got tests passing again. Anything else?

(If you don't respond for a bit, I'm assuming your approval was go-ahead to merge after fixing small notes)

tjdevries commented:
@bfredl you fine w/ this merge?

bfredl commented:
Yes.
